File: src/main/java/Lec8_Function.java
/**
 * Created by jihun.park@navercorp.com on 2022-10-04.
 */
public class Lec8_Function {

    public static void main(String[] args) {
        String[] strings = new String[]{"A","B","C"};
        printAll(strings);
        printAll("A","B","C");
    }

    public int max(int a, int b){
        if (a > b) {
            return a;
        }
        return b;
    }




    // repeat() - 주어진 문자열을 num 만큼 출력
    public void repeat(String str, int num, boolean useNewLine){
        for (int i = 0; i < num; i++) {
            if(useNewLine){
                System.out.println(str);
            }else{
                System.out.println(str);
            }
        }
    }
    // repeat() 오버로딩
    public void repeat(String str, int num){
        repeat(str,num,true);
    }
    // repeat() 오버로딩
    public void repeat(String str){
        repeat(str,3,true);
    }

    public static void printAll(String... strings) {
        for (String string : strings) {
            System.out.println("string = " + string);
        }
    }


}

File: src/main/kotlin/Lec8_Function.kt
/**
 * Created by jihun.park@navercorp.com on 2022-10-04.
 */

fun main(){
    repeat("Hello World", 3, true)

    // default parameter
    repeat("Hello World")

    // named argument
    repeat("Hello World", useNewLine = false)
    printNameAndGender("Jihun Park","M")
    printNameAndGender(gender = "M", name="Jihun Park") // builder 처럼 사용 가능

    // 가변인자
    val array = arrayOf("A","B","C")
    printAll(*array)
    printAll("A","B","C")
}

fun max(a: Int, b: Int): Int {
    return if (a > b) a else b // expression
}

fun max2(a: Int, b: Int) = if (a > b) a else b  // 결과가 하나일 경우 중괄호 생략

fun repeat(
    str: String,
    num: Int = 3,  // 기본값 설
    useNewLine: Boolean = true) { // 기본값 설정

    for(i in 1..num) {
        if (useNewLine) {
            println(str)
        } else {
            print(str)
        }
    }
}

fun printNameAndGender(name:String, gender: String){
    println(name)
    println(gender)
}

fun printAll(vararg strings:String){
    for (str in strings){
        print(str)
    }
}